src/cargo/core/features.rs

```diff
@@ -620,7 +620,30 @@ impl Features {
     ) -> CargoResult<()> {
         let nightly_features_allowed = self.nightly_features_allowed;
         let Some((slot, feature)) = self.status(feature_name) else {
-            bail!("unknown cargo feature `{}`", feature_name)
+            let mut msg = format!(
+                "feature `{feature_name}` is not a valid unstable feature for Cargo.toml\n\n"
+            );
+
+            if feature_name.contains('_') {
+                let _ = writeln!(msg, "Feature names must use '-' instead of '_'.");
+            } else {
+                let underscore_name = feature_name.replace('-', "_");
+                if CliUnstable::help()
+                    .iter()
+                    .any(|(option, _)| *option == underscore_name)
+                {
+                    let _ = writeln!(
+                        msg,
+                        "This feature can be enabled via -Z{feature_name} or the `[unstable]` section in config.toml."
+                    );
+                }
+            }
```

Comment:
No tests for these cases.  Please
- Make them end-to-end tests
- Add them in a commit before this one with them passing

Reply:
I've added tests.

Reply:
Could you please add them in the separate commit as asked

Reply:
What confused me was that the old behavior was already fully covered by tests, since it was so basic. So the separate test commit ends up being just copypaste of the same test three times.

Reply:
Yes, test commits on their own do not always provide value.  Its the diff between it and the follow up commit, along with the tests moving forward where the value is.